Update all followed tag with one click

kuei0221 opened the pull request:
<!--
     For Work In Progress Pull Requests, please use the Draft PR feature,
     see https://github.blog/2019-02-14-introducing-draft-pull-requests/ for further details.

     For a timely review/response, please avoid force-pushing additional
     commits if your PR already received reviews or comments.

     Before submitting a Pull Request, please ensure you've done the following:
     - 📖 Read the Forem Contributing Guide: https://github.com/forem/forem/blob/master/CONTRIBUTING.md#create-a-pull-request.
     - 📖 Read the Forem Code of Conduct: https://github.com/forem/forem/blob/master/CODE_OF_CONDUCT.md.
     - 👷‍♀️ Create small PRs. In most cases this will be possible.
     - ✅ Provide tests for your changes.
     - 📝 Use descriptive commit messages.
     - 📗 Update any related documentation and include any relevant screenshots.
-->

## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
Hi, this is my first PR for forem, so let me know if I miss anything 😄
This PR is for https://github.com/forem/forem/issues/1346, adding a new button to update all followed tags.
I follow an idea discussed in the issue:  remove all separated save button, and combine them into one.

By clicking the `Update Weight` button, an array of hash with id and weight point will be pass `[{ id:1, points: 3.0 }, { id: 2, points: 5.0 }]`, then the backend will handle the work. The form will only pass the changed weight point. If nothing changed yet, nothing pass.

Change in this PR:
1. Replace follows#update with follows#bulk_update in both route and controller
2. Replace FollowPolicy#update? with #bulk_update?
3. Combine all save button in following_tags.html.erb into one.
4. Mark changed weight points, and disable unchanged when submitting

## Related Tickets & Documents
https://github.com/forem/forem/issues/1346
## QA Instructions, Screenshots, Recordings
<img width="1377" alt="截圖 2020-11-26 上午2 17 19" src="https://user-images.githubusercontent.com/15964339/100266957-a2040a00-2f8d-11eb-8b03-a01cfb2e7aa5.png">


## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [ ] [Developer Docs](https://docs.forem.com) and/or [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [x] No documentation needed

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/forem/forem?pullRequest=11454) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/forem/forem?pullRequest=11454) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/forem/forem?pullRequest=11454) it.</sub>

mstruve reviewed (approved):
@kuei0221 welcome to the Forem community!!!! Thank you for taking the plunge and contributing to our project. We couldn't do what we do without awesome contributors like yourself. 

I did some manual testing and this works great! Backend looks solid to me, however, the js script code might be better off somewhere else. I tagged a couple of coworkers who have more experience with the frontend than myself to helps us figure that out 😉

Zhao-Andy commented on app/views/dashboards/following_tags.html.erb:
```diff
@@ -54,3 +60,36 @@
 </main>
+
+<script>
+  document.getElementById('follows_update_form').addEventListener('submit', checkChanged)
```
I think this script should live in our Webpack bundle, via `app/javascript/packs`. @kuei0221 I would put the JS code you have here in a regular JS file in that directory, and then you can load it on this `.erb` file like `
<%= javascript_packs_with_chunks_tag "newFileName", defer: true %>`.

Here's a bit more info on that process: https://docs.dev.to/frontend/webpacker/

kuei0221 replied:
Thanks! I will follow your guide 😄

Zhao-Andy commented on app/views/dashboards/following_tags.html.erb:
```diff
@@ -54,3 +60,36 @@
+    element.setAttribute('changed', true);
+  };
+
+  function checkChanged(event) {
+    if (document.querySelector('input[changed]')) {
+      disableAllUnchanged();
+    }
+    else {
+      event.preventDefault();
+    }
+  };
```
**non-blocking, nitpick**

I think this is a neat idea! I personally like having the button be `disabled` too if nothing is changed. Maybe you can add `button.disabled = true` in the `else`?

kuei0221 replied:
Thanks! But if we disable the button in the `else`, because the form submission is stopped by `preventDefault()`, all the buttons will be disabled and the user won't have access to change them until refreshing the page.

Zhao-Andy commented on app/views/dashboards/following_tags.html.erb:
```diff
@@ -54,3 +60,36 @@
+  };
+
+  function disableUnchanged(item) {
+    item = item.getElementsByTagName('input')
+    var id = item[0]
+    var point = item[1]
```
Ideally we use ES6 here and not use `var`. I think re-defining `item` to something more descriptive and not erasing the argument is best practice. :)

Zhao-Andy reviewed (commented):
Thanks for this PR! I think it's a great implementation so far :) Left some comments about a few details that I think should be cleaned up. Otherwise, it works well!

Zhao-Andy commented on app/assets/javascripts/initializers/initScrolling.js:
```diff
@@ -121,22 +121,13 @@ function buildTagsHTML(tag) {
     '</h3>' +
     '<p class="grid-cell__summary truncate-at-3"></p>' +
-    '<form class="edit_follow flex items-center flex-nowrap mb-4" id="edit_follow_' +
+    '<input name="follows[][id]" id="follow_id" type="hidden" form="follows_update_form" value="' +
```
I think `follows[id]` should work, don't need the extra empty bracket `[]`.

kuei0221 replied:
Rack will parse the empty bracket to an array (by [normalize_params](https://github.com/rack/rack/blob/1d70c73043e8fd41bf204ff794730aed2620d7a4/lib/rack/query_parser.rb#L102)), and append the following value to it.
So with `follows[][id]`, the params passed to the server will be `follows: [{id: 1}, {id: ...}, {...}]`
By doing this, the form can pass an array of hashes, which is easier for rails to handle the parameter restriction.

Here is a[ blog](http://codefol.io/posts/How-Does-Rack-Parse-Query-Params-With-parse-nested-query/) that describes this feature. Although this was post several years ago, rack remains the same mechanism.

I have fought on this trick for several hours, hope it makes sense 😅

Zhao-Andy commented on app/assets/javascripts/initializers/initScrolling.js:
```diff
@@ -121,22 +121,13 @@ function buildTagsHTML(tag) {
-    '" name="follow[points]" id="follow_points">' +
-    '<button type="submit" class="crayons-btn crayons-btn--ghost crayons-btn--s" name="commit">Save</button>' +
-    '</form>' +
+    '" name="follow[][points]" id="follow_points" onchange="addChanged(this)">' +
```
Also here -- this seems to work as `follow` or `follows` but let's be consistent and keep it as one or the other. :)

rhymes commented on app/controllers/follows_controller.rb:
```diff
@@ -79,16 +79,22 @@ def create
+  def bulk_update
+    @follows = Follow.where(id: params_for_update.keys).includes(:follower, :followable)
+    authorize @follows
+    redirect_to "/dashboard/following" if @follows.all? { |follow| follow.update(params_for_update[follow.id.to_s]) }
```
I wonder what would happen if any of this updates would fail. Should we separate the update step from the `if` and the redirection and do them one at the time?

ps. you can replace `/dashboard/following` with `dashboard_following_path`

kuei0221 replied:
Oops, thanks for reminding this. I think it would better to add a transaction for this and replace `update` with `update!`, so it will abort the entire update if any update step fails, so the process will be easy to trace. Thanks!

rhymes commented on app/views/dashboards/following_tags.html.erb:
```diff
@@ -19,6 +19,11 @@
         </div>
+
+        <%= form_with url: bulk_update_follows_path, method: :patch, local: true, id: 'follows_update_form' do |f| %>
+          <button type="submit" class="crayons-btn crayons-btn--secondary crayons-btn--m mb-3" name="commit">Update Weight</button>
```
should this be plural? Now that we update multiple weights

rhymes reviewed (commented):
Thanks @kuei0221, left a couple of notes. I'm also not convinced about the repeated `Follow weight:` labels, can we remove them? The page already has a help section explaining what the numbers are.

I would also change the "update weight" button to the main button, the blue one so it's more visible and it indicates the primary action

rhymes commented on app/javascript/packs/disableUnchangedButton.js:
```diff
@@ -0,0 +1,35 @@
+document.getElementById('follows_update_form').addEventListener('submit', checkChanged);
```
could you rename this file in something self explanatory?

I worry that `disableUnchangedButton` doesn't really communicate that it's related to the dashboard tags. Maybe something like `dashboardTagsDisableUnchangedButtons.js` ?

rhymes commented on app/javascript/packs/disableUnchangedButton.js:
```diff
@@ -0,0 +1,35 @@
+    point.setAttribute('disabled', true)
+    id.setAttribute('disabled', true)
+  }
+};
```
missing new line at the end of the file

rhymes commented on app/views/dashboards/following_tags.html.erb:
```diff
@@ -59,37 +60,4 @@
-  };
-
-</script>
+</main>
```
missing new line at the end of the file

rhymes commented on app/controllers/follows_controller.rb:
```diff
@@ -79,16 +79,25 @@ def create
+  def bulk_update
+    @follows = Follow.where(id: params_for_update.keys).includes(:follower, :followable)
+    authorize @follows
+    Follow.transaction do
```
should we write a test for this? Maybe by submitting a request with non decimal points? Something to make this explode

kuei0221 replied:
It is a great idea to add a test here, but I cannot figure out a request to make it explode.
ActiveRecord will change the type of data when saving (here points is float so is `to_f`), so nondecimal points or even negative integers will pass. Besides, the array or hash will not pass the policy before the transaction, and I didn't' find any validation for the weight point.

Do you have any idea? 🙏

rhymes replied:
That was the only one I had in mind. I don't think it's worth it to go down a rabbit hole of deep mocks. Can `follow.id` be invalid ?

kuei0221 replied:
I cannot find either. I have tried none exist id or updating the same id twice, but none works. None exist id will be ignored when `Follow.where`, duplicated id will only cut into one when `params_for_update`.

rhymes replied:
Ok, let's leave it like this in the meantime, I don't think it's super critical to have a test for the transaction that we know it's going to be rolled back. If you could address the other notes we can merge this afterwards. Thanks for bouncing ideas!

kuei0221 commented:
I have updated the code according to the new merged PR #11513, renamed the point which can be updated by user to `explicit_points`.
Now the tag insert by `initScrolling.js` will still show calculated points to the user (but update explicit_points), because the inserted tag is generated by another controller action `followings#tags` and it needs an update. I think it a new issue and needs another PR for this.
And maybe it will be nice if we have a test to check what generated by `initScrolling` is equal to the origin. 😉

rhymes commented:
@benhalpern should this functionality use `explicit_points` directly ?

Zhao-Andy reviewed (approved):
Looks great to me! Thanks so much for the PR! Excited for people to use this feature :)

Zhao-Andy commented:
Sorry, I dismissed my approval because I think we should use `explicit_points` based on the description from #11513. Will wait for @benhalpern to confirm.

citizen428 commented on app/controllers/follows_controller.rb:
```diff
@@ -79,17 +79,25 @@ def create
-    authorize @follow
-    @follow.explicit_points = follow_params[:explicit_points]
-    redirect_to dashboard_following_path if @follow.save
+  def bulk_update
```
**suggestion (non-blocking):** This is due to my personal dislike for non-REsTful actions in Rails controllers, but I'd probably do a nested `BulkUpdatesController` with a resourceful route declaration in `routes.rb`, see http://jeromedalbert.com/how-dhh-organizes-his-rails-controllers/

nickytonline commented:
From a UX standpoint, it would probably make sense to make the *Update Weights* button sticky. Otherwise, so save, the user would always have to scroll back to the top. cc: @ludwiczakpawel

ludwiczakpawel commented:
> From a UX standpoint, it would probably make sense to make the _Update Weights_ button sticky. Otherwise, so save, the user would always have to scroll back to the top. cc: @ludwiczakpawel

Yes, definitely.

In fact, we should probably put that button in the bottom, below the grid. And once any value changes, make it sticky <-- we do exactly that when you try to update something in https://dev.to/settings - the "Save" button sticks to the bottom edge of the screen.

Would be great to ship it that way, but if this is too much of a hassle we can probably do it in a separate PR.

benhalpern commented:
Sorry for the delay, yes @Zhao-Andy @rhymes we should use `explicit_points` here.

`explicit_points` are the ones the user interacts with directly, `implicit_points` are the ones that the system interacts with to nudge based on user activity, and `points` is the sum of these values which are ultimately used for sorting in general.

rhymes reviewed (approved):
Ben confirmed, re-tested this manually, LGTM!

Thanks @kuei0221!

@nickytonline @ludwiczakpawel I think we can work on improving the UX in a separate PR as this has been open for quite a while

nickytonline commented:
> Ben confirmed, re-tested this manually, LGTM!
> 
> Thanks @kuei0221!
> 
> @nickytonline @ludwiczakpawel I think we can work on improving the UX in a separate PR as this has been open for quite a while

Done. I created a new issue.
